type-metadata/src/lib.rs

```diff
@@ -1,7 +1,338 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		assert_eq!(2 + 2, 4);
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(CustomIdent),
+
+	// primitives and common preludes
+	Slice(SliceIdent),
+	Array(ArrayIdent),
+	Tuple(TupleIdent),
+	Bool,
+	Str,
+	U8,
+	U16,
+	U32,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	I64,
+	I128,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct Namespace {
+	segments: Vec<&'static str>,
+}
+impl Namespace {
+	pub fn new(segments: Vec<&'static str>) -> Self {
+		Namespace { segments }
+	}
+	fn prelude() -> Self {
+		Namespace { segments: vec![] }
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct CustomIdent {
+	pub name: &'static str,
+	pub namespace: Namespace,
+	pub type_params: Vec<IdentKind>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ArrayIdent {
+	pub len: u16,
+	pub type_param: Box<IdentKind>,
+}
+impl ArrayIdent {
+	fn new(len: u16, type_param: IdentKind) -> Self {
+		ArrayIdent {
+			len,
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TupleIdent {
+	pub type_params: Vec<IdentKind>,
+}
+impl TupleIdent {
+	fn new(type_params: Vec<IdentKind>) -> Self {
+		TupleIdent { type_params }
+	}
+	fn unit() -> Self {
+		TupleIdent { type_params: vec![] }
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct SliceIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl SliceIdent {
+	fn new(type_param: IdentKind) -> Self {
+		SliceIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, Debug)]
```

Comment:
What I had in mind was something like the following the would be structurally closely aligned to Rust's type structure:
```rust
#[derive(Debug, PartialEq, Eq, Serialize)]
pub enum MetaTypeKind {
    NamedStruct(NamedStruct),
    TupleStruct(TupleStruct),
    Enum(MetaTypeEnum),
    Variant(MetaTypeVariant),
    Union(MetaTypeUnion),
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct NamedStruct {
    fields: Vec<NamedField>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct NamedField {
    name: &'static str,
    #[serde(rename = "type")]
    ty: TypeInfoId,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct TupleStruct {
    fields: Vec<UnnamedField>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct UnnamedField {
    index: u32,
    #[serde(rename = "type")]
    ty: TypeInfoId,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct ClikeEnum {
    variants: Vec<EnumCase>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct ClikeEnumVariant {
    name: &'static str,
    index: u32,
    default_value: Option<u64>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct Enum {
    variants: Vec<Variant>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub enum EnumVariant {
    Unit(VariantUnit),
    Struct(VariantStruct),
    Tuple(VariantTuple),
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct VariantUnit {
    name: &'static str,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct VariantStruct {
    fields: Vec<NamedField>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct VariantTuple {
    fields: Vec<UnnamedField>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct Union {
    fields: Vec<UnionField>,
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct UnionField {
    name: String,
    #[serde(rename = "type")]
    ty: TypeInfoId,
}
```

Note that `TypeInfoId` is basically what is `TypeIdent` in your PR and `MetaTypeKind` is similar to `TypeDef`.

Reply:
And: We really want/need some examples and tests that make use of `TypeDef` in order to check if it has a decent structure etc.

Reply:
I cannot see the benefit of separating enum to `ClikeEnum` and `Enum`. `ClikeEnum` is more like an edge case when all variants are C-Like. As there is no limitation of mixing C-Like variants and data variants in an enum, maybe it's better to only keep `Enum` with a C-like variant:

```rust
#[derive(Debug, PartialEq, Eq, Serialize)]
pub enum Variant {
    Clike(VariantClike),
    Unit(VariantUnit),
    Struct(VariantStruct),
    Tuple(VariantTuple),
}
```